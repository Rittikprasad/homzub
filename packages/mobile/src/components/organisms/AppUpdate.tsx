/**
 * Handling Force Update, Maintenance logic using remote config
 * Doc for ref: https://rnfirebase.io/remote-config/usage
 */

import React, { useEffect, useState } from 'react';
import { Linking, StyleSheet, View } from 'react-native';
import remoteConfig from '@react-native-firebase/remote-config';
import { AppModes, ConfigHelper, ConfigModes } from '@homzhub/common/src/utils/ConfigHelper';
import { DeviceUtils } from '@homzhub/common/src/utils/DeviceUtils';
import { Logger } from '@homzhub/common/src/utils/Logger';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { I18nService } from '@homzhub/common/src/services/Localization/i18nextService';
import Rocket from '@homzhub/common/src/assets/images/rocket.svg';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { BottomSheet } from '@homzhub/common/src/components/molecules/BottomSheet';

// Remote config Parameters
enum ConfigKeys {
  MINIMUM_VERSION = 'minimum_version',
  LATEST_VERSION = 'latest_version',
  IS_UNDER_MAINTENANCE = 'is_under_maintenance',
  MESSAGES = 'messages',
}

interface IMessages {
  update: string;
  maintenance: string;
}

interface IConfig {
  [key: string]: any;
}

interface IOwnState {
  isForceUpdate: boolean;
  isUpdateAvailable: boolean;
  isUnderMaintenance: boolean;
  messages: IMessages;
}

const initialState: IOwnState = {
  isForceUpdate: false,
  isUpdateAvailable: false,
  isUnderMaintenance: false,
  messages: {
    update: '',
    maintenance: '',
  },
};

const AppUpdate = (): React.ReactElement => {
  const [configData, setConfigData] = useState<IConfig[]>([]);
  const [configValues, setConfigValues] = useState<IOwnState>(initialState);
  const [isSheetVisible, setVisibility] = useState(false);

  useEffect(() => {
    remoteConfig()
      .setConfigSettings({
        minimumFetchIntervalMillis: 30000, // Global Caching frequency
      })
      .then(() => {
        // Set Default before fetching config from firebase
        remoteConfig()
          .setDefaults({
            is_under_maintenance: false,
            latest_version: '',
            minimum_version: '',
          })
          .then(() => remoteConfig().fetchAndActivate()) // Fetch config and Activate
          .then(() => {
            const config = remoteConfig().getAll(); // Get All config list

            if (config) {
              const formattedData = Object.entries(config).map((item) => {
                const [key, entry] = item as any;
                return {
                  [key]: JSON.parse(entry.asString()),
                };
              });
              setConfigData(formattedData);
            }
          })
          .then(() => remoteConfig().fetch()); // Fetching data for next process
      })

      .catch((e) => {
        Logger.warn(e.message);
      });
  }, []);

  useEffect(() => {
    const updatedData = { ...configValues };
    const configMode = ConfigHelper.getConfigMode();
    const appMode = (configMode === ConfigModes.PROD ? AppModes.RELEASE : AppModes.DEBUG).toLowerCase();

    // Structuring config and maintaining local state to handle UI
    configData.forEach((item) => {
      if (Object.keys(item)[0] === ConfigKeys.MINIMUM_VERSION) {
        updatedData.isForceUpdate = item[ConfigKeys.MINIMUM_VERSION][appMode] > DeviceUtils.appVersion;
      } else if (Object.keys(item)[0] === ConfigKeys.LATEST_VERSION) {
        updatedData.isUpdateAvailable = item[ConfigKeys.LATEST_VERSION][appMode] > DeviceUtils.appVersion;
      } else if (Object.keys(item)[0] === ConfigKeys.IS_UNDER_MAINTENANCE) {
        updatedData.isUnderMaintenance = item[ConfigKeys.IS_UNDER_MAINTENANCE][appMode];
      } else {
        updatedData.messages = item[ConfigKeys.MESSAGES];
      }
    });

    const isVisible = updatedData.isUpdateAvailable || updatedData.isUnderMaintenance || updatedData.isForceUpdate;

    setVisibility(isVisible);
    setConfigValues(updatedData);
  }, [configData]);

  const onUpdateNow = async (): Promise<void> => {
    const { getAppleStoreUrl, getGooglePlayStoreUrl } = ConfigHelper;
    await Linking.openURL(PlatformUtils.isIOS() ? getAppleStoreUrl() : getGooglePlayStoreUrl());
  };

  const { isForceUpdate, messages, isUnderMaintenance, isUpdateAvailable } = configValues;
  const isAlwaysVisible = isForceUpdate || isUnderMaintenance;
  return (
    <BottomSheet
      visible={isSheetVisible}
      isShadowView
      headerTitle={isUpdateAvailable ? I18nService.t('common:timeToUpdate') : I18nService.t('common:underMaintenance')}
      sheetHeight={isUpdateAvailable ? 420 : 300}
      isCloseOnDrag={!isAlwaysVisible}
      isAlwaysVisible={isAlwaysVisible}
      onCloseSheet={(): void => setVisibility(false)}
    >
      <View style={styles.container}>
        <Rocket style={styles.iconStyle} />
        <Label type="large" style={styles.message}>
          {isUpdateAvailable ? messages?.update : messages?.maintenance}
        </Label>
        {!isUnderMaintenance && (
          <Button
            type="primary"
            title={I18nService.t('common:updateNow')}
            containerStyle={styles.updateButton}
            onPress={onUpdateNow}
          />
        )}
        {!isAlwaysVisible && (
          <Button
            type="secondaryOutline"
            title={I18nService.t('common:doLater')}
            containerStyle={styles.secondaryButton}
            textStyle={styles.buttonTitle}
            onPress={(): void => setVisibility(false)}
          />
        )}
      </View>
    </BottomSheet>
  );
};

export default AppUpdate;

const styles = StyleSheet.create({
  container: {
    padding: 32,
    flex: 1,
  },
  iconStyle: {
    alignSelf: 'center',
  },
  message: {
    textAlign: 'center',
    marginVertical: 16,
    color: theme.colors.darkTint3,
  },
  updateButton: {
    flex: 0,
    height: 46,
    marginVertical: 10,
  },
  secondaryButton: {
    height: 46,
    flex: 0,
  },
  buttonTitle: {
    color: theme.colors.darkTint3,
  },
});
