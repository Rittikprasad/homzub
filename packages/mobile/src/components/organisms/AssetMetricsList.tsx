import React, { useState, useCallback } from 'react';
import { StyleProp, StyleSheet, TouchableOpacity, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useRoute } from '@react-navigation/native';
import { useNavigation } from '@react-navigation/core';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import Logo from '@homzhub/common/src/assets/images/homzhubDashboard.svg';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { PaginationComponent } from '@homzhub/mobile/src/components/atoms/PaginationComponent';
import { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';
import { AssetMetrics } from '@homzhub/mobile/src/components/molecules/AssetMetrics';

export interface IMetricsData {
  name: string;
  count: string | number;
  label?: string;
  id?: number;
  isCurrency?: boolean;
  colorCode: string;
  code?: string;
}

interface IProps {
  data: IMetricsData[];
  title?: string;
  subscription?: string;
  selectedAssetType?: string;
  numOfElements?: number;
  isSubTextRequired?: boolean;
  onPlusIconClicked?: () => void;
  onMetricsClicked?: (name: string) => void;
  containerStyle?: StyleProp<ViewStyle>;
  subTitleText?: string;
  headerIcon?: string;
  showBackIcon?: boolean;
}

const COMPONENT_PADDING = 12;
const SLIDER_WIDTH = theme.viewport.width - theme.layout.screenPadding * 2;
const AssetMetricsList = (props: IProps): React.ReactElement => {
  const {
    title = 0,
    data,
    selectedAssetType,
    onPlusIconClicked,
    containerStyle,
    onMetricsClicked,
    numOfElements = 3,
    isSubTextRequired = true,
    subTitleText,
    headerIcon,
    showBackIcon = false,
  } = props;

  // HOOKS
  const { t } = useTranslation();
  const { goBack, setParams } = useNavigation();
  const { params } = useRoute();
  const [activeIndex, setActiveIndex] = useState(0);
  // HELPERS
  const bubblePlusIcon = useCallback((): void => {
    if (onPlusIconClicked) {
      onPlusIconClicked();
    }
  }, [onPlusIconClicked]);

  const splitData = useCallback((): IMetricsData[][] => {
    const newArr = [];

    for (let i = 0; i < data.length; i += numOfElements) {
      newArr.push(data.slice(i, i + numOfElements));
    }

    return newArr;
  }, [data, numOfElements]);

  const onGoBack = (): void => {
    goBack();
    // @ts-ignore
    if (params && params?.isFromNavigation) {
      setParams({ isFromNavigation: false });
    }
  };
  // HELPERS END

  const renderItem = useCallback(
    (items: IMetricsData[]): React.ReactElement => {
      return (
        <View style={styles.sliderRow}>
          {items.map((item: IMetricsData) => {
            const cardStyle = {
              minWidth: (SLIDER_WIDTH - COMPONENT_PADDING * (numOfElements + 1)) / numOfElements,
            };
            // TODO: (Shikha) - Connect with BE to handle all logic with CODE
            const handlePress = (): void => onMetricsClicked && onMetricsClicked(item.code || item.name);
            return (
              <AssetMetrics
                key={`${item.label ?? item.name}`}
                header={item.label ?? item.name}
                value={item.count}
                isCurrency={item.isCurrency ?? false}
                cardStyle={cardStyle}
                colorCode={item.colorCode}
                onPressMetrics={handlePress}
                selectedAssetType={selectedAssetType}
                // Todo : Check in multiple devices
                maxLength={items.length === 2 ? 15 : undefined}
              />
            );
          })}
        </View>
      );
    },
    [onMetricsClicked, numOfElements, selectedAssetType]
  );

  return (
    <View style={[styles.container, containerStyle]}>
      <View style={[styles.header, !isSubTextRequired && styles.financialView]}>
        {showBackIcon && (
          <TouchableOpacity onPress={onGoBack} style={styles.backIcon}>
            <Icon name={icons.leftArrow} size={25} color={theme.colors.blue} />
          </TouchableOpacity>
        )}
        {headerIcon ? (
          <View>
            <Icon name={headerIcon} size={30} style={styles.icon} color={theme.colors.blue} />
            <View style={[styles.headerIcon, styles.logo]} />
          </View>
        ) : (
          <Logo style={styles.logo} width={50} height={50} />
        )}
        <View style={styles.headerCenter}>
          <Text type="regular" textType="bold" style={styles.assetCount}>
            {title}
          </Text>
          <View style={styles.propertiesRow}>
            {isSubTextRequired && (
              <Label type="large" textType="semiBold" style={styles.propertyText}>
                {subTitleText || t('common:properties')}
              </Label>
            )}
          </View>
        </View>
        {onPlusIconClicked && (
          <TouchableOpacity onPress={bubblePlusIcon} style={styles.plusIconContainer}>
            <Icon name={icons.circularPlus} size={32} color={theme.colors.primaryColor} testID="icnPlus" />
          </TouchableOpacity>
        )}
      </View>
      <SnapCarousel
        carouselData={splitData()}
        carouselItem={renderItem}
        activeIndex={activeIndex}
        onSnapToItem={setActiveIndex}
        sliderWidth={SLIDER_WIDTH}
        itemWidth={SLIDER_WIDTH}
      />
      <PaginationComponent
        dotsLength={splitData().length}
        activeSlide={activeIndex}
        containerStyle={styles.paginationContainer}
        activeDotStyle={[styles.dot, styles.activeDot]}
        inactiveDotStyle={[styles.dot, styles.inactiveDot]}
      />
    </View>
  );
};

const memoizedComponent = React.memo(AssetMetricsList);
export { memoizedComponent as AssetMetricsList };

const styles = StyleSheet.create({
  container: {
    paddingVertical: 12,
    borderRadius: 4,
    backgroundColor: theme.colors.white,
  },
  header: {
    marginHorizontal: 12,
    flexDirection: 'row',
  },
  headerCenter: {
    flex: 1,
  },
  propertiesRow: {
    flexDirection: 'row',
  },
  sliderRow: {
    flexDirection: 'row',
  },
  paginationContainer: {
    paddingVertical: 0,
    marginTop: 12,
  },
  assetCount: {
    color: theme.colors.darkTint1,
  },
  propertyText: {
    color: theme.colors.darkTint4,
  },
  logo: {
    marginEnd: 12,
  },
  dot: {
    width: 8.5,
    height: 8.5,
  },
  activeDot: {
    borderWidth: 1.5,
  },
  inactiveDot: {
    backgroundColor: theme.colors.disabled,
    borderWidth: 0,
  },
  financialView: {
    alignItems: 'center',
  },
  headerIcon: {
    backgroundColor: theme.colors.blue,
    width: 50,
    height: 50,
    opacity: 0.1,
    borderRadius: 30,
  },
  icon: {
    position: 'absolute',
    left: 10,
    top: 10,
  },
  plusIconContainer: {
    justifyContent: 'center',
    padding: 5,
  },
  backIcon: {
    alignSelf: 'center',
    marginRight: 10,
  },
});
