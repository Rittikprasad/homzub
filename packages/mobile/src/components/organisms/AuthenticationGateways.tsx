import React from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { WithTranslation, withTranslation } from 'react-i18next';
import { StackNavigationProp } from '@react-navigation/stack';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { SocialAuthUtils } from '@homzhub/common/src/utils/SocialAuthUtils';
import { AuthService } from '@homzhub/mobile/src/services/AuthService';
import { ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import Apple from '@homzhub/common/src/assets/images/apple.svg';
import Google from '@homzhub/common/src/assets/images/google.svg';
import Facebook from '@homzhub/common/src/assets/images/facebook.svg';
import LinkedIn from '@homzhub/common/src/assets/images/linkedin.svg';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import CommonAuthenticationHoc from '@homzhub/common/src/components/organisms/CommonAuthenticationHOC';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { SocialAuthProvider } from '@homzhub/common/src/domain/models/SocialAuthProvider';
import { ISocialUserData, SocialAuthKeys } from '@homzhub/common/src/constants/SocialAuthProviders';
import { IUserTokens } from '@homzhub/common/src/services/storage/StorageService';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IDispatchProps {
  loginSuccess: (data: IUserTokens) => void;
}

interface IAuthenticationGatewayProps {
  isFromLogin: boolean;
  onEmailLogin?: () => void;
  testID?: string;
  navigation?: StackNavigationProp<AuthStackParamList, ScreensKeys.SignUp | ScreensKeys.Login>;
  toggleLoading?: (isLoading: boolean) => void;
  onSuccessCallback?: () => void;
  isSocialAllowed?: boolean;
}

type Props = IAuthenticationGatewayProps & WithTranslation & IDispatchProps;

interface IOwnState {
  authProviders: SocialAuthProvider[];
}

class AuthenticationGateways extends React.PureComponent<Props, IOwnState> {
  public state = {
    authProviders: [],
  };

  public componentDidMount(): void {
    SocialAuthUtils.fetchSocialMedia((response) => {
      if (PlatformUtils.isIOS()) {
        response.unshift(
          ObjectMapper.deserialize(SocialAuthProvider, {
            provider: SocialAuthKeys.Apple,
            description: 'Apple',
            visible: PlatformUtils.isIOS(),
          })
        );
      }

      this.setState({ authProviders: response });
    });
  }

  public render(): React.ReactNode {
    const { onEmailLogin, isFromLogin, isSocialAllowed = true } = this.props;

    return (
      <CommonAuthenticationHoc isFromLogin={isFromLogin} onEmailLogin={onEmailLogin}>
        {!PlatformUtils.isIOS() && isSocialAllowed && this.renderSocialAuths()}
      </CommonAuthenticationHoc>
    );
  }

  private renderSocialAuths = (): React.ReactNode => {
    const { authProviders } = this.state;

    return authProviders.map((auth: SocialAuthProvider) => {
      const initiateSocialLogin = async (): Promise<void> => {
        await this.initiateSocialAuthentication(auth.provider);
      };

      return (
        <>
          {auth.visible && (
            <TouchableOpacity style={styles.alignToCenter} onPress={initiateSocialLogin} key={auth.provider}>
              {this.getImageUrlForLib(auth.provider)}
              <Label type="regular" textType="regular" style={styles.iconTextStyle}>
                {auth.description}
              </Label>
            </TouchableOpacity>
          )}
        </>
      );
    });
  };

  private onSocialAuthSuccess = async (userData: ISocialUserData): Promise<void> => {
    const { loginSuccess, navigation, isFromLogin, onSuccessCallback } = this.props;

    if (navigation) {
      await SocialAuthUtils.onSocialAuthSuccess(
        userData,
        () => {
          navigation.navigate(ScreensKeys.MobileVerification, {
            isFromLogin,
            userData,
          });
        },
        (response) => {
          loginSuccess(response);
          if (onSuccessCallback) {
            onSuccessCallback();
          }
        }
      );
    }
  };

  private initiateSocialAuthentication = async (key: string): Promise<void> => {
    const { isFromLogin, navigation } = this.props;
    this.setLoading(true);
    if (key === SocialAuthKeys.Google) {
      await AuthService.signInWithGoogle(this.onSocialAuthSuccess);
      this.setLoading(false);
      return;
    }

    if (key === SocialAuthKeys.Apple) {
      await AuthService.signInWithApple(this.onSocialAuthSuccess);
      this.setLoading(false);
      return;
    }

    if (key === SocialAuthKeys.Facebook) {
      await AuthService.signInWithFacebook(this.onSocialAuthSuccess, isFromLogin && navigation);
      this.setLoading(false);
    }
  };

  private setLoading = (isLoading: boolean): void => {
    const { toggleLoading } = this.props;
    if (toggleLoading) {
      toggleLoading(isLoading);
    }
  };

  private getImageUrlForLib = (key: string): React.ReactNode => {
    switch (key) {
      case SocialAuthKeys.Facebook:
        return <Facebook />;
      case SocialAuthKeys.Google:
        return <Google />;
      case SocialAuthKeys.LinkedIn:
        return <LinkedIn />;
      case SocialAuthKeys.Apple:
        return <Apple />;
      default:
        return null;
    }
  };
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { loginSuccess } = UserActions;
  return bindActionCreators(
    {
      loginSuccess,
    },
    dispatch
  );
};

const HOC = connect(
  null,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.auth)(AuthenticationGateways));
export { HOC as AuthenticationGateways };

const styles = StyleSheet.create({
  iconTextStyle: {
    marginTop: 4,
  },
  alignToCenter: {
    alignItems: 'center',
  },
});
