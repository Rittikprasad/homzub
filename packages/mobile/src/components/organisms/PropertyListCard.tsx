import React from 'react';
import { StyleProp, StyleSheet, View, ViewStyle, TouchableOpacity, Image } from 'react-native';
import { WithTranslation, withTranslation } from 'react-i18next';
import { PropertyUtils } from '@homzhub/common/src/utils/PropertyUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Divider } from '@homzhub/common/src/components/atoms/Divider';
import { Favorite } from '@homzhub/common/src/components/atoms/Favorite';
import { ImagePlaceholder } from '@homzhub/common/src/components/atoms/ImagePlaceholder';
import { PricePerUnit } from '@homzhub/common/src/components/atoms/PricePerUnit';
import { PropertyAddress } from '@homzhub/common/src/components/molecules/PropertyAddress';
import { PropertyAmenities } from '@homzhub/common/src/components/molecules/PropertyAmenities';
import { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';
import { ShieldGroup } from '@homzhub/mobile/src/components/molecules/ShieldGroup';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { Attachment } from '@homzhub/common/src/domain/models/Attachment';
import { TextSizeType } from '@homzhub/common/src/components/atoms/Text';
import { IAmenitiesIcons } from '@homzhub/common/src/domain/models/Search';

interface IProps {
  property: Asset;
  transaction_type: number;
  containerStyle?: StyleProp<ViewStyle>;
  isCarousel: boolean;
  textSizeType?: TextSizeType;
  onSelectedProperty: () => void;
  testID?: string;
  isAssetOwner: boolean;
}

type libraryProps = WithTranslation;
type Props = libraryProps & IProps;

interface IState {
  ref: any;
  activeSlide: number;
}

const ITEM_WIDTH = theme.viewport.width - 56;
export class PropertyListCard extends React.PureComponent<Props, IState> {
  public state = {
    activeSlide: 0,
    ref: null,
  };

  public render(): React.ReactElement {
    const {
      property: { projectName, unitNumber, blockNumber, address },
      containerStyle,
      onSelectedProperty,
    } = this.props;

    return (
      <View style={[styles.container, containerStyle]}>
        {this.renderCarousel()}
        {this.renderPropertyTypeAndBadges()}
        <TouchableOpacity onPress={onSelectedProperty}>
          <PropertyAddress
            isIcon
            primaryAddress={projectName}
            subAddress={address || `${blockNumber ?? ''} ${unitNumber ?? ''}`}
          />
          <Divider containerStyles={styles.divider} />
          {this.renderPriceAndAmenities()}
        </TouchableOpacity>
      </View>
    );
  }

  private renderPropertyTypeAndBadges = (): React.ReactElement => {
    const {
      property: {
        assetType: { name },
      },
    } = this.props;
    return <ShieldGroup propertyType={name} />;
  };

  private renderCarousel = (): React.ReactNode => {
    const { activeSlide } = this.state;
    const {
      isCarousel,
      property: { leaseTerm, saleTerm, sortedImages },
      isAssetOwner,
    } = this.props;

    return (
      <View style={styles.carouselContainer}>
        {!isCarousel ? (
          this.renderCarouselItem(sortedImages[0])
        ) : sortedImages.length > 0 ? (
          <SnapCarousel
            bubbleRef={this.updateRef}
            carouselData={sortedImages}
            carouselItem={this.renderCarouselItem}
            activeIndex={activeSlide}
            itemWidth={ITEM_WIDTH}
            sliderWidth={ITEM_WIDTH}
            onSnapToItem={this.onSnapToItem}
            testID="snapCarousel"
          />
        ) : (
          <ImagePlaceholder height="100%" />
        )}
        <View style={styles.overlay}>
          <View style={styles.favoriteContainer}>
            {!isAssetOwner && <Favorite leaseId={leaseTerm?.id} saleId={saleTerm?.id} iconColor={theme.colors.white} />}
          </View>
          {isCarousel && (
            <View style={styles.arrowContainer}>
              <Icon
                name={icons.leftArrow}
                size={25}
                color={activeSlide === 0 ? theme.colors.disabledPreference : theme.colors.white}
                onPress={this.previousSlide}
              />
              <Icon
                name={icons.rightArrow}
                size={25}
                color={activeSlide === sortedImages.length - 1 ? theme.colors.disabledPreference : theme.colors.white}
                onPress={this.nextSlide}
              />
            </View>
          )}
        </View>
      </View>
    );
  };

  private renderPriceAndAmenities = (): React.ReactElement => {
    const { transaction_type, property, textSizeType = 'regular' } = this.props;
    const {
      carpetArea,
      carpetAreaUnit,
      spaces,
      furnishing,
      country: { currencies },
      assetGroup: { code },
      saleTerm,
      leaseTerm,
    } = property;
    const price: number = this.getPrice();
    const amenitiesData: IAmenitiesIcons[] = PropertyUtils.getAmenities(
      spaces,
      furnishing,
      code,
      carpetArea,
      carpetAreaUnit?.title ?? ''
    );
    let currencyData = currencies[0];

    if (leaseTerm && leaseTerm.currency) {
      currencyData = leaseTerm.currency;
    }

    if (saleTerm && saleTerm.currency) {
      currencyData = saleTerm.currency;
    }

    return (
      <View style={styles.amenities}>
        <PricePerUnit
          price={price}
          currency={currencyData}
          unit={transaction_type === 0 ? 'mo' : ''}
          textSizeType={textSizeType}
        />
        <PropertyAmenities data={amenitiesData} direction="row" contentContainerStyle={styles.amenitiesContainer} />
      </View>
    );
  };

  private renderCarouselItem = (item: Attachment): React.ReactElement => {
    if (!item) {
      return <ImagePlaceholder height="100%" />;
    }

    return (
      <Image
        source={{
          uri: item.link,
        }}
        style={styles.carouselImage}
        resizeMode="contain"
      />
    );
  };

  private onSnapToItem = (slideNumber: number): void => {
    this.setState({ activeSlide: slideNumber });
  };

  private updateRef = (ref: any): void => {
    this.setState({ ref });
  };

  private previousSlide = (): void => {
    const { ref } = this.state;
    // @ts-ignore
    ref.snapToPrev();
  };

  private nextSlide = (): void => {
    const { ref } = this.state;
    // @ts-ignore
    ref.snapToNext();
  };

  private getPrice = (): number => {
    const {
      property: { leaseTerm, saleTerm },
    } = this.props;
    if (leaseTerm) {
      return Number(leaseTerm.expectedPrice);
    }
    if (saleTerm) {
      return Number(saleTerm.expectedPrice);
    }
    return 0;
  };
}

export default withTranslation()(PropertyListCard);

const styles = StyleSheet.create({
  container: {
    backgroundColor: theme.colors.white,
    padding: 12,
    borderRadius: 4,
    marginVertical: 10,
  },
  divider: {
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: theme.colors.disabled,
    marginVertical: 6,
  },
  amenities: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  amenitiesContainer: {
    marginTop: 8,
  },
  carouselContainer: {
    borderRadius: 4,
    height: 210,
    overflow: 'hidden',
  },
  arrowContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    margin: 8,
    marginVertical: 40,
  },
  overlay: {
    position: 'absolute',
    flex: 1,
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: theme.colors.carouselCardOpacity,
  },
  carouselImage: {
    height: '100%',
    width: '100%',
  },
  favoriteContainer: {
    flexDirection: 'row-reverse',
    margin: 15,
  },
});
