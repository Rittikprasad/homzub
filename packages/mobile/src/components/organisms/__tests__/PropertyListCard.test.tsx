import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyListCard } from '@homzhub/mobile/src/components/organisms/PropertyListCard';
import { mockAsset } from '@homzhub/common/src/mocks/AssetDescription';

let props: any;
let wrapper: ShallowWrapper;

describe.skip('PropertyListCard', () => {
  const createTestProps = (testProps: any): object => ({
    property: mockAsset,
    onFavorite: jest.fn(),
    transaction_type: 1,
    onSelectedProperty: jest.fn(),
    isCarousel: true,
    ...testProps,
  });

  props = createTestProps({});

  it('should match snapshot', () => {
    wrapper = shallow(<PropertyListCard {...props} t={(key: string): string => key} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot for optional props', () => {
    props = createTestProps({
      transaction_type: 0,
    });
    wrapper = shallow(<PropertyListCard {...props} t={(key: string): string => key} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
