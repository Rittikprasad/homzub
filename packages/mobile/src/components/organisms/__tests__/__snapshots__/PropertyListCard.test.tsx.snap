// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyListCard should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 4,
        "marginVertical": 10,
        "padding": 10,
      },
      undefined,
    ]
  }
>
  <PropertyListImageCarousel
    images={
      Array [
        Object {
          "file_name": "prof.jpg",
          "is_cover_image": true,
          "link": "https://homzhub-bucket.s3.amazonaws.com/asset_images/8e8c48fc-c089-11ea-8247-34e12d38d70eprof.jpg",
          "media_attributes": Object {},
          "media_type": "IMAGE",
        },
        Object {
          "file_name": "DjangoBackendLLD (1).jpg",
          "is_cover_image": false,
          "link": "https://homzhub-bucket.s3.amazonaws.com/asset_images/01d71336-cd92-11ea-972b-34e12d38d70eDjangoBackendLLD%201.jpg",
          "media_attributes": Object {},
          "media_type": "IMAGE",
        },
      ]
    }
    isFavorite={false}
    onFavorite={[MockFunction]}
  />
  <ShieldGroup
    propertyType="Apartment / Condo"
  />
  <ForwardRef
    onPress={[MockFunction]}
  >
    <PropertyAddress
      isIcon={true}
      subAddress=" "
    />
    <Divider
      containerStyles={
        Object {
          "borderColor": "#C9D6DF",
          "borderWidth": 0.5,
          "marginVertical": 6,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginVertical": 6,
          "overflow": "hidden",
        }
      }
    >
      <Memo(PricePerUnit)
        currency={
          Object {
            "currency_code": "",
            "currency_name": "",
            "currency_symbol": "",
          }
        }
        price={NaN}
        textSizeType="regular"
        unit=""
      />
      <PropertyAmenities
        contentContainerStyle={
          Object {
            "marginStart": 5,
          }
        }
        data={
          Array [
            Object {
              "icon": "furnishing",
              "iconColor": "#33495E",
              "iconSize": 20,
              "label": "None",
            },
            Object {
              "icon": "open-parking",
              "iconColor": "#33495E",
              "iconSize": 20,
              "label": "-",
            },
            Object {
              "icon": "area",
              "iconColor": "#33495E",
              "iconSize": 20,
              "label": "-",
            },
          ]
        }
        direction="row"
      />
    </View>
  </ForwardRef>
</View>
`;

exports[`PropertyListCard should match snapshot for optional props 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 4,
        "marginVertical": 10,
        "padding": 10,
      },
      undefined,
    ]
  }
>
  <PropertyListImageCarousel
    images={
      Array [
        Object {
          "file_name": "prof.jpg",
          "is_cover_image": true,
          "link": "https://homzhub-bucket.s3.amazonaws.com/asset_images/8e8c48fc-c089-11ea-8247-34e12d38d70eprof.jpg",
          "media_attributes": Object {},
          "media_type": "IMAGE",
        },
        Object {
          "file_name": "DjangoBackendLLD (1).jpg",
          "is_cover_image": false,
          "link": "https://homzhub-bucket.s3.amazonaws.com/asset_images/01d71336-cd92-11ea-972b-34e12d38d70eDjangoBackendLLD%201.jpg",
          "media_attributes": Object {},
          "media_type": "IMAGE",
        },
      ]
    }
    isFavorite={false}
    onFavorite={[MockFunction]}
  />
  <ShieldGroup
    propertyType="Apartment / Condo"
  />
  <ForwardRef
    onPress={[MockFunction]}
  >
    <PropertyAddress
      isIcon={true}
      subAddress=" "
    />
    <Divider
      containerStyles={
        Object {
          "borderColor": "#C9D6DF",
          "borderWidth": 0.5,
          "marginVertical": 6,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginVertical": 6,
          "overflow": "hidden",
        }
      }
    >
      <Memo(PricePerUnit)
        currency={
          Object {
            "currency_code": "",
            "currency_name": "",
            "currency_symbol": "",
          }
        }
        price={NaN}
        textSizeType="regular"
        unit="mo"
      />
      <PropertyAmenities
        contentContainerStyle={
          Object {
            "marginStart": 5,
          }
        }
        data={
          Array [
            Object {
              "icon": "furnishing",
              "iconColor": "#33495E",
              "iconSize": 20,
              "label": "None",
            },
            Object {
              "icon": "open-parking",
              "iconColor": "#33495E",
              "iconSize": 20,
              "label": "-",
            },
            Object {
              "icon": "area",
              "iconColor": "#33495E",
              "iconSize": 20,
              "label": "-",
            },
          ]
        }
        direction="row"
      />
    </View>
  </ForwardRef>
</View>
`;
